app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/MainActivity.kt

```diff
@@ -16,14 +16,60 @@
 
 package com.raulh82vlc.functionalkotlinarrow.ui
 
-import android.support.v7.app.AppCompatActivity
 import android.os.Bundle
+import android.support.v7.app.AppCompatActivity
+import android.support.v7.widget.LinearLayoutManager
+import android.view.View
+import android.widget.Toast
+import arrow.effects.fix
 import com.raulh82vlc.functionalkotlinarrow.R
+import com.raulh82vlc.functionalkotlinarrow.data.cache.model.FeedItemCacheModel
+import com.raulh82vlc.functionalkotlinarrow.presentation.FeedListView
+import com.raulh82vlc.functionalkotlinarrow.presentation.Presentation
+import com.raulh82vlc.functionalkotlinarrow.ui.adapter.FeedAdapter
+import kotlinx.android.synthetic.main.activity_main.*
 
-class MainActivity : AppCompatActivity() {
+class MainActivity : AppCompatActivity(), FeedListView {
+    private lateinit var feedAdapter: FeedAdapter
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        setFeedList()
+
+        Presentation.showFeed(this).fix().unsafeRunAsync {}
+    }
+
+    private fun setFeedList() {
+        feed_rv.setHasFixedSize(true)
+        feed_rv.layoutManager = LinearLayoutManager(this)
+        feedAdapter = FeedAdapter(itemClick = {
+            Presentation.onClickFeedItem(this, it.authorId).fix().unsafeRunAsync {}
+        })
+        feed_rv.adapter = feedAdapter
+    }
+
+    override fun showNetworkError(t: Throwable) = runOnUiThread {
+        Toast.makeText(this, t.localizedMessage, Toast.LENGTH_LONG).show()
+        showNoResults()
+    }
+
+    override fun showFeed(feed: List<FeedItemCacheModel>) = runOnUiThread {
+        if (feed.isEmpty()) {
+            showNoResults()
+        } else {
+            showResults()
+            feedAdapter.fillData(feed)
+        }
+    }
+
+    private fun showNoResults() {
+        no_results_view.visibility = View.VISIBLE
+        feed_rv.visibility = View.GONE
+    }
+
+    private fun showResults() {
+        no_results_view.visibility = View.GONE
+        feed_rv.visibility = View.VISIBLE
     }
```

Comment:
This can be modelled with a sealed class `State` type:
```
sealed class State {
  object Empty : State()
  data class Failed(val message: String) : State()
  data class Ready(val feed: List<FeedItemCacheModel>) : State()
}
```

Reply:
good one, seems legit

Reply:
will do for a new refactor though